tools/CodeGen/StronglyTypedOverloadsGenerator.cs

```diff
@@ -0,0 +1,87 @@
+namespace CodeGen;
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+public abstract class StronglyTypedOverloadsGenerator : ISourceGenerator
+{
+    protected const int MaximumTypeParameterCount = 8;
+
+    private static readonly string[] Ordinals =
+    [
+        "zeroth",
+        "first",
+        "second",
+        "third",
+        "fourth",
+        "fifth",
+        "sixth",
+        "seventh",
+        "eighth"
+    ];
+
+    public void Execute(GeneratorExecutionContext context)
+    {
+        var source = $"""
+            //------------------------------------------------------------------------------
+            // <auto-generated>
+            //     This code was generated by {this.GetType().FullName}.
+            //
+            //     Changes to this file will be lost when the project is rebuilt.
+            // </auto-generated>
+            //------------------------------------------------------------------------------
+            {this.GenerateSource()}
+            """;
+        context.AddSource(
+            this.GetType().Name.Replace("Generator", ".StronglyTyped.cs"),
+            source);
+    }
+
+    public void Initialize(GeneratorInitializationContext context)
+    {
+        // No initialization required
+    }
+
+    protected static string OverloadMethod(Func<int, string> generator) =>
+        Escalate(MaximumTypeParameterCount, "\n\n", generator);
+
+    protected static string TypeParamDocsSource(int typeParametersCount) =>
+        Escalate(
+            typeParametersCount,
+            "\n",
+            i => $"""/// <typeparam name="T{i}">The type of the {Ordinals[i]} argument of the faked method call.</typeparam>""");
+
+    protected static string TypeParamList(int typeParametersCount) =>
+        Escalate(typeParametersCount, ", ", TypeParamName);
+
+    /// <summary>
+    /// Generates a string by invoking the <paramref name="generator" /> function for each number
+    /// from 1 to <paramref name="maximum" /> and joining the results
+    /// with <paramref name="separator"/>.
+    /// </summary>
+    /// <param name="maximum">The maximum number to generate strings for.</param>
+    /// <param name="separator">The separator to use between generated strings.</param>
+    /// <param name="generator">The function to generate strings.</param>
+    /// <returns>The generated string.</returns>
+    protected static string Escalate(int maximum, string separator, Func<int, string> generator) =>

```

Comment:
I don't _love_ the name of this, but struggled.

Reply:
Yeah, I didn't immediately understand what it did when I saw the call. But I don't have a better idea. It's fine, I think.